### PR TITLE
Femdem/minor changes pfemdem

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm correcting some thing regarding the pfem-femdem coupling. I'm also adding the NeoHookean cl from structural (will be removed when the porting to the SMA will be done). A minor change in the DEM App `dem_fem_coupling_algorithm.py` (@maceligueta )

### CODE REVIEW

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 1]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
             it_node->Fix(VELOCITY_X);
             it_node->Fix(VELOCITY_Y);
             it_node->Fix(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                auto current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 2]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                auto prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 3]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                auto &r_current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
+                auto &r_prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_PREVIOUS_BACKUP);
+                r_current_accel_backup = current_acceleration;
+                r_prev_accel_backup    = prev_acceleration;
+
+                auto current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 4]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                auto prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 5]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
             it_node->Fix(VELOCITY_Y);
             it_node->Fix(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                auto current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
+                auto prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
+                auto &r_current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
+                auto &r_prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_PREVIOUS_BACKUP);
+                r_current_accel_backup = current_acceleration;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_current_accel_backup) = current_acceleration;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 6]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_prev_accel_backup    = prev_acceleration;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_prev_accel_backup)  = prev_acceleration;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 7]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+
+                auto current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
+                auto prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
+                auto &r_current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP);
+                auto &r_prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_PREVIOUS_BACKUP);
+                r_current_displ_backup = current_displ;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_current_displ_backup) = current_displ;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 8]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_prev_displ_backup    = prev_displ;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_prev_displ_backup) = prev_displ;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 9]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
         } else {
             it_node->Free(VELOCITY_X);
             it_node->Free(VELOCITY_Y);
             it_node->Free(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                auto &r_current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
+                auto &r_prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
+                auto current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
```

Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 10]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                auto prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_PREVIOUS_BACKUP);
+                r_current_acceleration = current_accel_backup;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_current_acceleration) = current_accel_backup;
```

Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_current_acceleration) = current_accel_backup;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 11]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_prev_acceleration = prev_accel_backup;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_prev_acceleration) = prev_accel_backup;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 12]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+
+                auto &r_current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
+                auto &r_prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
+                auto current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]: This backups take a lot of memory, and why do you need to have to variables for backup?, all historical variables have buffer, you can acces from previous positions

File: applications/FemToDemApplication/custom_python/fem_to_dem_python_application.cpp [hunk 13]
```diff
@@ -61,6 +61,10 @@ namespace Python
     AddCustomProcessesToPython(m);
 
     //registering variables in python
+    KRATOS_REGISTER_IN_PYTHON_3D_VARIABLE_WITH_COMPONENTS(m,ACCELERATION_BACKUP);
+    KRATOS_REGISTER_IN_PYTHON_3D_VARIABLE_WITH_COMPONENTS(m,ACCELERATION_PREVIOUS_BACKUP);
+    KRATOS_REGISTER_IN_PYTHON_3D_VARIABLE_WITH_COMPONENTS(m,DISPLACEMENT_BACKUP);
+    KRATOS_REGISTER_IN_PYTHON_3D_VARIABLE_WITH_COMPONENTS(m,DISPLACEMENT_PREVIOUS_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_python/fem_to_dem_python_application.cpp]: Remove this please and use the buffer from DISPLACEMENT_BACKUP, idem for ACCELERATION_BACKUP

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 14]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
             it_node->Fix(VELOCITY_X);
             it_node->Fix(VELOCITY_Y);
             it_node->Fix(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                auto current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
+                auto prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
+                auto &r_current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
+                auto &r_prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_PREVIOUS_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                auto &r_prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 15]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_current_accel_backup = current_acceleration;
+                r_prev_accel_backup    = prev_acceleration;
+
+                auto current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
+                auto prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
+                auto &r_current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP);
+                auto &r_prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_PREVIOUS_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                auto &r_prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 16]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
             it_node->Free(VELOCITY_X);
             it_node->Free(VELOCITY_Y);
             it_node->Free(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                auto &r_current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
+                auto &r_prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
+                auto current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP);
+                auto prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_PREVIOUS_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                auto prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 17]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_current_acceleration = current_accel_backup;
+                r_prev_acceleration = prev_accel_backup;
+
+                auto &r_current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
+                auto &r_prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
+                auto current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP);
+                auto prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_PREVIOUS_BACKUP);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                conts auto& prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 18]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_current_displ = current_displ_backup;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_current_displ) = current_displ_backup;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 19]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                r_prev_displ = prev_displ_backup;
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                noalias(r_prev_displ) = prev_displ_backup;
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp (hunk 12 above)
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                const auto& current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp (hunk 14 above)
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp]:
```suggestion
                auto &r_prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 1);
```

File: applications/FemToDemApplication/fem_to_dem_application.cpp [hunk 20]
```diff
@@ -59,6 +59,10 @@ void KratosFemToDemApplication::Register()
      KratosApplication::Register();
     
     //REGISTER VARIABLES FEM2DEM
+    KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_BACKUP)
+    KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_PREVIOUS_BACKUP)
+    KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(DISPLACEMENT_BACKUP)
+    KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(DISPLACEMENT_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/fem_to_dem_application.cpp]: Remove

File: applications/FemToDemApplication/fem_to_dem_application.cpp [hunk 21]
```diff
@@ -59,6 +59,10 @@ void KratosFemToDemApplication::Register()
      KratosApplication::Register();
     
     //REGISTER VARIABLES FEM2DEM
+    KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_BACKUP)
+    KRATOS_REGISTER_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/fem_to_dem_application.cpp]: Remove

File: applications/FemToDemApplication/fem_to_dem_application_variables.cpp [hunk 22]
```diff
@@ -14,6 +14,10 @@
 
 namespace Kratos
 {
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_BACKUP)
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/fem_to_dem_application_variables.cpp]: Remove

File: applications/FemToDemApplication/fem_to_dem_application_variables.cpp [hunk 23]
```diff
@@ -14,6 +14,10 @@
 
 namespace Kratos
 {
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_BACKUP)
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_PREVIOUS_BACKUP)
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(DISPLACEMENT_BACKUP)
+    KRATOS_CREATE_3D_VARIABLE_WITH_COMPONENTS(DISPLACEMENT_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/fem_to_dem_application_variables.cpp]: Remove

File: applications/FemToDemApplication/fem_to_dem_application_variables.h [hunk 24]
```diff
@@ -25,6 +25,10 @@
 
 namespace Kratos
 {
+    KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_BACKUP)
+    KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/fem_to_dem_application_variables.h]: Remove

File: applications/FemToDemApplication/fem_to_dem_application_variables.h [hunk 25]
```diff
@@ -25,6 +25,10 @@
 
 namespace Kratos
 {
+    KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_BACKUP)
+    KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(ACCELERATION_PREVIOUS_BACKUP)
+    KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(DISPLACEMENT_BACKUP)
+    KRATOS_DEFINE_3D_VARIABLE_WITH_COMPONENTS(DISPLACEMENT_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/fem_to_dem_application_variables.h]: Remove

File: applications/FemToDemApplication/python_scripts/MainFEM_for_PFEM_coupling.py [hunk 26]
```diff
@@ -80,7 +80,10 @@ def Initialize(self):
         self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PRESSURE_VELOCITY)
         self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PRESSURE_REACTION)
         self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PRESSURE_ACCELERATION)
-
+        self.main_model_part.AddNodalSolutionStepVariable(KratosFemDem.ACCELERATION_BACKUP)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosFemDem.ACCELERATION_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/python_scripts/MainFEM_for_PFEM_coupling.py]: Remove

File: applications/FemToDemApplication/python_scripts/MainFEM_for_PFEM_coupling.py [hunk 27]
```diff
@@ -80,7 +80,10 @@ def Initialize(self):
         self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PRESSURE_VELOCITY)
         self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PRESSURE_REACTION)
         self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PRESSURE_ACCELERATION)
-
+        self.main_model_part.AddNodalSolutionStepVariable(KratosFemDem.ACCELERATION_BACKUP)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosFemDem.ACCELERATION_PREVIOUS_BACKUP)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosFemDem.DISPLACEMENT_BACKUP)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosFemDem.DISPLACEMENT_PREVIOUS_BACKUP)
```
Comment (loumalouomega) [inline applications/FemToDemApplication/python_scripts/MainFEM_for_PFEM_coupling.py]: Remove

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Many things to be removed, mainly the unnecessary duplicated backup when you have to your disposal the buffer

**Comment by AlejandroCornejo** [comment]
Dear @loumalouomega , the backups are done because of the PFEM-FEMDEM coupling. Summarizing, the pfem sets to 0 the accelerations of the boundaries (FEMDEM skin) and this is a problem for the FEMDEM solve. With the backups I can save the acc and reset them after the PFEM solves.

In my opinion has nothing to do with the buffer :)

Thank you kindly,

**Comment by loumalouomega** [comment]
> 
> 
> Dear @loumalouomega , the backups are done because of the PFEM-FEMDEM coupling. Summarizing, the pfem sets to 0 the accelerations of the boundaries (FEMDEM skin) and this is a problem for the FEMDEM solve. With the backups I can save the acc and reset them after the PFEM solves.
> 
> In my opinion has nothing to do with the buffer :)
> 
> Thank you kindly,

If your buffer is size 2, and you want  to save 2 values, you are actually saving 4 values!!! If you use the second position of the buffer as I suggested you will not waste half of memory

**Comment by AlejandroCornejo** [comment]
> 
> 
> > Dear @loumalouomega , the backups are done because of the PFEM-FEMDEM coupling. Summarizing, the pfem sets to 0 the accelerations of the boundaries (FEMDEM skin) and this is a problem for the FEMDEM solve. With the backups I can save the acc and reset them after the PFEM solves.
> > In my opinion has nothing to do with the buffer :)
> > Thank you kindly,
> 
> If your buffer is size 2, and you want to save 2 values, you are actually saving 4 values!!! If you use the second position of the buffer as I suggested you will not waste half of memory

so you're saying that I only have to create one new variables and use the buffer, okkey :) changing it

**Comment by AlejandroCornejo** [comment]
At least I have to create a new backup because the PFEM app resets the current acc and the previous...

**Comment by loumalouomega** [comment]
> 
> 
> At least I have to create a new backup because the PFEM app resets the current acc and the previous...

Yeah, but my question is why do it in two different variables and not only one variable

**Comment by AlejandroCornejo** [comment]
> 
> 
> > At least I have to create a new backup because the PFEM app resets the current acc and the previous...
> 
> Yeah, but my question is why do it in two different variables and not only one variable

because I need the current and the previous acceleration and the same with the displ... :)

**Comment by loumalouomega** [comment]
> 
> 
> > > Dear @loumalouomega , the backups are done because of the PFEM-FEMDEM coupling. Summarizing, the pfem sets to 0 the accelerations of the boundaries (FEMDEM skin) and this is a problem for the FEMDEM solve. With the backups I can save the acc and reset them after the PFEM solves.
> > > In my opinion has nothing to do with the buffer :)
> > > Thank you kindly,
> > 
> > 
> > If your buffer is size 2, and you want to save 2 values, you are actually saving 4 values!!! If you use the second position of the buffer as I suggested you will not waste half of memory
> 
> so you're saying that I only have to create one new variables and use the buffer, okkey :) changing it

Yes I meant that, is already changed?

**Comment by AlejandroCornejo** [comment]
yes @loumalouomega I've changed that

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 28]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
         } else {
             it_node->Free(VELOCITY_X);
             it_node->Free(VELOCITY_Y);
             it_node->Free(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                auto &r_current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
+                auto &r_prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
+                const auto current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 0);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:69]:
```suggestion
                const auto& current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 0);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 29]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                const auto prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 1);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:70]:
```suggestion
                const auto& prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 30]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                noalias(r_current_acceleration) = current_accel_backup;
+                noalias(r_prev_acceleration) = prev_accel_backup;
+
+                auto &r_current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
+                auto &r_prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
+                const auto current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP, 0);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:76]:
```suggestion
                const auto& current_displ_backup = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP, 0);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 31]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                const auto prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP, 1);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:77]:
```suggestion
                const auto& prev_displ_backup    = it_node->FastGetSolutionStepValue(DISPLACEMENT_BACKUP, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 32]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
             it_node->Fix(VELOCITY_X);
             it_node->Fix(VELOCITY_Y);
             it_node->Fix(VELOCITY_Z);
+
+            // We store previous info in order to reset after PFEM Solve
+            if (r_process_info[STEP] > 1) { // We have the previous acc
+                const auto current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:45]:
```suggestion
                const auto& current_acceleration = it_node->FastGetSolutionStepValue(ACCELERATION, 0);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 33]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                const auto prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:46]:
```suggestion
                const auto& prev_acceleration    = it_node->FastGetSolutionStepValue(ACCELERATION, 1);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 34]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                auto &r_current_accel_backup = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 0);
+                auto &r_prev_accel_backup    = it_node->FastGetSolutionStepValue(ACCELERATION_BACKUP, 1);
+                noalias(r_current_accel_backup) = current_acceleration;
+                r_prev_accel_backup    = prev_acceleration;
+
+                const auto current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:52]:
```suggestion
                const auto& current_displ = it_node->FastGetSolutionStepValue(DISPLACEMENT, 0);
```

File: applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp [hunk 35]
```diff
@@ -38,10 +39,45 @@ void FixFreeVelocityOnNodesProcess::Execute()
...
+                const auto prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/fix_free_velocity_on_nodes_process.cpp:53]:
```suggestion
                const auto& prev_displ    = it_node->FastGetSolutionStepValue(DISPLACEMENT, 1);
```

**Review by loumalouomega — COMMENTED** [review]
Some missing &

**Comment by AlejandroCornejo** [comment]
> 
> 
> Some missing &

I tried to use the referendcesw but there was memory corruption so... I prefer to leave it as it is now :)

**Comment by loumalouomega** [comment]
> 
> 
> > Some missing &
> 
> I tried to use the referendcesw but there was memory corruption so... I prefer to leave it as it is now :)

Even whit the noalias now?, I think that with the noalias should copy the result not reference it. Can you try?

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**